packages/expo-ota/android/src/main/java/expo/modules/ota/ExpoOTAConfig.kt

```diff
@@ -0,0 +1,85 @@
+        val manifestResponseValidator: ManifestResponseValidator
+)
+val defaultManifestComparator: ManifestComparator = VersionNumberManifestComparator(SdkVersionMustBeEqualNativeComparator())
+private fun defaultValidator(manifestClient: OkHttpClient) = ExpoValidator("https://exp.host/--/manifest-public-key", manifestClient)
+
+fun embeddedManifestExpoConfig(context: Context): ExpoOTAConfig {
+    val embeddedManifestAndBundle = EmbeddedManifestAndBundle(context).readManifest()
+    val releaseChannel = embeddedManifestAndBundle.getString("releaseChannel")
+    val sdkVersion = embeddedManifestAndBundle.getString("sdkVersion")
+    val hostUri = embeddedManifestAndBundle.getString("hostUri")
+    val updatesConfig: JSONObject? = embeddedManifestAndBundle.optJSONObject("updates")
+    var checkAutomatically = defaultCheckAutomatically
+    var manifestComparator = defaultManifestComparator
+    if(updatesConfig != null) {
+        checkAutomatically = updatesConfig.optString("checkAutomatically") != "ON_ERROR_RECOVERY"
+        val manifestComparatorValue = updatesConfig.optString("versionComparison")
+        manifestComparator = manifestComparatorByVersionComparisonValue(manifestComparatorValue)
+    }
+    val manifestClient = OkHttpClient()
+    return ExpoOTAConfig(manifestUrl = "https://$hostUri",
+            manifestHeaders = manifestHeaders(expoSdkVersion = sdkVersion, releaseChannel = releaseChannel),
+            channelIdentifier = releaseChannel,
+            manifestComparator = manifestComparator,
+            manifestHttpClient = manifestClient,
+            checkForUpdatesAutomatically = checkAutomatically,
+            manifestResponseValidator = defaultValidator(manifestClient))
+}
+
+@JvmOverloads
+@Suppress("unused")
+fun expoHostedOTAConfig(username: String,
+                        projectName: String,
+                        releaseChannel: String = "default",
+                        validateManifestResponse: Boolean = true,
+                        expoSdkVersion: String = "35.0.0",
+                        manifestHttpClient: OkHttpClient? = null,
+                        bundleHttpClient: OkHttpClient? = null,
+                        manifestComparator: ManifestComparator = defaultManifestComparator,
+                        apiVersion: Number = 1): ExpoOTAConfig {
+    val manifestClient = manifestHttpClient
+            ?: OkHttpClient()
+    return ExpoOTAConfig(manifestUrl = "https://exp.host/@$username/$projectName",
+            manifestHeaders = manifestHeaders(expoSdkVersion, apiVersion, releaseChannel),
+            manifestHttpClient = manifestClient,
+            channelIdentifier = releaseChannel,
+            manifestComparator = manifestComparator,
+            bundleHttpClient = bundleHttpClient,
+            manifestResponseValidator = if (validateManifestResponse) ExpoValidator("https://exp.host/--/manifest-public-key", manifestClient) else DummyValidator())
```

Comment:
One thing that's really important is that the manifest OkHttpClient should never read from a cache. Currently in ExpoKit we use `CacheControl.FORCE_NETWORK` to enforce this. Otherwise we might end up missing an update, since the manifest url, in theory, never changes.

On the other hand, for the validator we have pretty specific logic to use a cache for the public key and then only retry with network if the validation fails (in case we rotated the keys). Would be nice to replicate that logic if possible. https://github.com/expo/expo/blob/380493569893c76fe905c3295bcbfb3cc3bb3865/android/expoview/src/main/java/host/exp/exponent/kernel/Crypto.java#L52-L56

Reply:
I believe that for the cache to be used, server would need to response with `304` response code. It might change once we set `Cache-Control` to 'Force-Cache`, but as long as there is no forcing, it should work automatically right. And as far as I was testing it, this is the case. Even for very frequent updates (like within 10 seconds), none of updates were missing.

When it comes to caching public key, I believe that we can reimplement that behavior.